### PR TITLE
[Clang] Emit TBAA info for enums in C

### PR DESCRIPTION
**Pull request by david-arm**
When emitting TBAA information for enums in C code we currently just treat the data as an 'omnipotent char'. However, with C strict aliasing this means we fail to optimise certain cases. For example, in the SPEC2017 xz benchmark there are structs that contain arrays of enums, and clang pessmistically assumes that accesses to those enums could alias with other struct members that have a different type.

According to

https://en.cppreference.com/w/c/language/enum

enums should be treated as 'int' types unless explicitly specified (C23) or if 'int' would not be large enough to hold all the enumerated values. In the latter case the compiler is free to choose a suitable integer that would hold all such values.

When compiling C code this patch generates TBAA information for the enum by using an equivalent integer of the size clang has already chosen for the enum. I have ignored C++ for now because the rules are more complex.

New test added here:

  clang/test/CodeGen/tbaa.c

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenTBAA.cpp [hunk 1]
```diff
@@ -196,11 +196,14 @@ llvm::MDNode *CodeGenTBAA::getTypeInfoHelper(const Type *Ty) {
   // Enum types are distinct types. In C++ they have "underlying types",
   // however they aren't related for TBAA.
   if (const EnumType *ETy = dyn_cast<EnumType>(Ty)) {
+    if (!Features.CPlusPlus)
+      return getTypeInfo(Context.getIntTypeForBitwidth(Size * 8, 0));
```
Comment (david-arm) [inline clang/lib/CodeGen/CodeGenTBAA.cpp:200]: I am not sure if this is entirely correct so would appreciate some guidance here!

Comment (momchil-velikov) [reply clang/lib/CodeGen/CodeGenTBAA.cpp:200]: Maybe `ETy->getDecl()->getIntegerType()` ?

Comment (david-arm) [reply clang/lib/CodeGen/CodeGenTBAA.cpp]: Good suggestion - thanks!

**Comment by david-arm** [comment]
Gentle ping!

**Comment by AaronBallman** [comment]
> enums should be treated as 'int' types unless explicitly specified (C23) or if 'int' would not be large enough to hold all the enumerated values. In the latter case the compiler is free to choose a suitable integer that would hold all such values.

In C23, the enumeration type is compatible with its underlying type, which is either `char` or a standard or extended signed or unsigned integer type (C23 6.7.2.2p2). The enumeration member type (C23 6.7.2.2p11) is specified by p12 (when building the enumeration itself) and p15 and p16 (once the enumeration is complete). If there's no fixed underlying type, the members are compatible with either `int` or the enumerated type, and if there is a fixed underlying type, the members type is the enumerated type. Much of this is implementation-defined behavior. It sounds like our implementation was assuming compatibility with `char` and not `int` because of the aliasing behavior, and now you'd like to change it to be `int` so we don't have the aliasing problems?

**Comment by david-arm** [comment]
Hi @AaronBallman, yes the problem I found with always choosing `char` as the alias type is that LLVM will just assume that enum types alias with absolutely everything. This is a conservative approach that works fine, but it does prevent important type-based alias optimisations from happening. GCC seems to take advantage of the fact that enums without any explicit type should be compatible with an `int` (or a suitable integer that the compiler is free to choose that can contain all enumerated values). Consequently it can be more aggressive in optimisations, which is why I started looking at the C specification to understand the rules. If my understanding is correct, I think the important thing is that the type-based alias information should match the actual underlying type we've chosen for the enum. It could be, for example that in the case of an enum like this:

```
enum Foo {
  A = 1,
  B = 0x100000000ull
};
```

gcc chooses a `long long` and clang chooses a `long` since there seems to be wiggle room in the specification. I think that's fine provided the compiler is self-consistent. In either case both `long` and `long long` would not alias with `int` or `short`, which is still an improvement on treating it as a `char` in TBAA info.

**Comment by efriedma-quic** [comment]
The C notion of "compatibility" with regard to enums is a bit weird... clang and gcc disagree whether `enum e : int; enum e2: int; enum e x; int x; enum e2 x;` is valid because 6.2.7 doesn't actually define what's supposed to happen.  Probably this should be illegal, but the standard isn't really clear here.  Maybe it should be clarified to have a notion of "redeclaration compatibility" which is separate from general type compatibility.

That said, this patch is pretty conservative; it's hard for me to imagine treating an enum as the underlying type could cause any issues (unless code is assuming an enum has universal aliasing).

I guess this should be compatible with older clang versions, so we don't need to rev the version of TBAA data.

**Comment by rjmccall** [comment]
This seems conservatively correct, yeah.  My reading is that we could also use the underlying type as a parent type for the TBAA metadata: enums are compatible with their underlying type, but two enums with the same underlying type are not compatible with each other.  But this rule should also work.

**Comment by paulwalker-arm** [comment]
Do you think it's worth adding something to the Clang release note?

**Comment by AaronBallman** [comment]
> Do you think it's worth adding something to the Clang release note?

Yes, please!

**Comment by momchil-velikov** [comment]
I thought the suggestion was to add a few lines to https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst

**Comment by david-arm** [comment]
> I thought the suggestion was to add a few lines to https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst

Yes you're right! For some reason I got mixed up with the LangRef, but I guess adding something to the LangRef does no harm either. I'll put something in the release notes too. :)

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -6376,13 +6376,26 @@ aliases a memory access with an access tag ``(BaseTy2, AccessTy2,
 Offset2)`` if either ``(BaseTy1, Offset1)`` is reachable from ``(Base2,
 Offset2)`` via the ``Parent`` relation or vice versa.
 
+In C an enum will be compatible with an underlying integer type that is large
+enough to hold all enumerated values. In most cases this will be an ``int``,
+which is the default when no type is specified. However, if an ``int`` is not
```
Comment (paulwalker-arm) [inline llvm/docs/LangRef.rst:6381]: I don't think this information belongs in the LangRef, especially as it doesn't add any value to the underlying description of what the metadata represents.

**Review by paulwalker-arm — APPROVED** [review]
This looks broadly good to me.  I suggest reverting the LangRef change because it doesn't add any new information relevant to LLVM IR.

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -196,6 +196,9 @@ C Language Changes
   number of elements in the flexible array member. This information can improve
   the results of the array bound sanitizer and the
   ``__builtin_dynamic_object_size`` builtin.
+- Enums will now be represented in TBAA metadata using their actual underlying
```
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply clang/docs/ReleaseNotes.rst:199]: Works for me.

**Pull request merged**